src/mmap_unix.rs

```diff
@@ -74,6 +74,152 @@ impl error::Error for Error {}
+    pub fn with_bitmap(mut self, bitmap: B) -> Self {
+        self.bitmap = Some(bitmap);
+        self
+    }
+
+    /// Create the `MmapRegion` object with pre-mmapped raw pointer.
+    ///
+    /// # Safety
+    ///
+    /// To use this safely, the caller must guarantee that `raw_addr` and `self.size` define a
+    /// region within a valid mapping that is already present in the process.
+    pub unsafe fn with_raw_mmap_pointer(mut self, raw_ptr: *mut u8) -> Self {
+        self.raw_ptr = Some(raw_ptr);
+        self
+    }
+
+    /// Build the `MmapRegion` object.
+    pub fn build(mut self) -> Result<MmapRegion<B>> {
+        if self.raw_ptr.is_some() {
+            return self.build_raw();
+        }
+
+        // Forbid MAP_FIXED, as it doesn't make sense in this context, and is pretty dangerous
+        // in general.
+        if self.flags & libc::MAP_FIXED != 0 {
+            return Err(Error::MapFixed);
+        }
+
+        let (fd, offset) = if let Some(ref f_off) = self.file_offset {
+            check_file_offset(f_off, self.size)?;
+            (f_off.file().as_raw_fd(), f_off.start())
+        } else {
+            (-1, 0)
+        };
+
+        // This is safe because we're not allowing MAP_FIXED, and invalid parameters cannot break
+        // Rust safety guarantees (things may change if we're mapping /dev/mem or some wacky file).
+        let addr = unsafe {
+            libc::mmap(
+                null_mut(),
+                self.size,
+                self.prot,
+                self.flags,
+                fd,
+                offset as libc::off_t,
+            )
+        };
+
+        if addr == libc::MAP_FAILED {
+            return Err(Error::Mmap(io::Error::last_os_error()));
+        }
+
+        let bitmap = self.bitmap.take().unwrap_or_else(|| B::with_len(self.size));
+
+        Ok(MmapRegion {
+            addr: addr as *mut u8,
+            size: self.size,
+            bitmap,
+            file_offset: self.file_offset,
+            prot: self.prot,
+            flags: self.flags,
+            owned: true,
+            hugetlbfs: self.hugetlbfs,
+        })
+    }
+
+    fn build_raw(mut self) -> Result<MmapRegion<B>> {
+        // Safe because this call just returns the page size and doesn't have any side effects.
+        let page_size = unsafe { libc::sysconf(libc::_SC_PAGESIZE) } as usize;
+        let addr = self.raw_ptr.clone().unwrap();
+
+        // Check that the pointer to the mapping is page-aligned.
+        if (addr as usize) & (page_size - 1) != 0 {
+            return Err(Error::InvalidPointer);
+        }
+
+        // Check that the size is a multiple of the page size.
+        if self.size & (page_size - 1) != 0 {
+            return Err(Error::InvalidSize);
+        }
+
+        let bitmap = self.bitmap.take().unwrap_or_else(|| B::with_len(self.size));
+
+        Ok(MmapRegion {
+            addr,
+            size: self.size,
+            bitmap,
+            file_offset: self.file_offset,
```

Comment:
Can `file_offset` be `Some` here?

Reply:
It's possible with new builder api.

Reply:
So if the user provides a file_offset and a raw_ptr when building `MmapRegionBuilder`, file_offset is basically ignored. Should `file_offset` be None here then? :-?

Reply:
Even with  build_raw(), user still could attach an file object to the MmapRegion object and retrieve the attached file by MmapRegion::file_offset(). This way it will be helpful for the user to make use of build_raw().